### ISSUE TITLE
Expose a `TypedResults` method to push a partial response

### ISSUE THREAD

**Issue by ranma42** [dotnet/aspnetcore#55753]
## Background and Motivation

In a project I am developing I have the need to accept ranged requests and provide the client with partial responses (206).
The data for these responses is managed in a chunk-based store which is somewhat awkward to wrap in a `Stream`.
For full responses, the `PushStreamHttpResult` is ideal, but it currently does not support partial requests.
It is quite straightforward to implement an `IResult` that does what I need, but it looks like I would need to re-implement `HttpResultsHelper.WriteResultAsFileCore` as it is an `internal` method.

## Proposed API

```diff
namespace Microsoft.AspNetCore.Http;

public static class TypedResults
{
+    /// <summary>
+    /// Allows writing directly to the response body.
+    /// <para>
+    /// This supports range requests (<see cref="StatusCodes.Status206PartialContent"/> or
+    /// <see cref="StatusCodes.Status416RangeNotSatisfiable"/> if the range is not satisfiable).
+    /// </para>
+    /// </summary>
+    /// <param name="callback">The callback that allows users to write directly to the HTTP response.</param>
+    /// <param name="fileLength">The total length of the file.</param>
+    /// <param name="contentType">The <c>Content-Type</c> of the response. Defaults to <c>application/octet-stream</c>.</param>
+    /// <param name="fileDownloadName">The the file name to be used in the <c>Content-Disposition</c> header.</param>
+    /// <param name="lastModified">The <see cref="DateTimeOffset"/> of when the file was last modified.
+    /// Used to configure the <c>Last-Modified</c> response header and perform conditional range requests.</param>
+    /// <param name="entityTag">The <see cref="EntityTagHeaderValue"/> to be configure the <c>ETag</c> response header
+    /// and perform conditional requests.</param>
+    /// <param name="enableRangeProcessing">Set to <c>true</c> to enable range requests processing.</param>
+    /// <returns>The created <see cref="PushStreamHttpResult"/> for the response.</returns>
+    public static PushStreamHttpResult Stream(
+        Func<RangeItemHeaderValue?, HttpContext, Task> callback,
+        long? fileLength = null,
+        string? contentType = null,
+        string? fileDownloadName = null,
+        DateTimeOffset? lastModified = null,
+        EntityTagHeaderValue? entityTag = null,
+        bool enableRangeProcessing = false);
```

## Usage Examples

```csharp
// respond with a stream representing the selected part of a resource
// this avoids moving around any useless data
app.MapGet("/example1/{key}", async (string key, IStore store) =>
    TypedResults.Stream(
        async (range, httpContext) => {
            using var stream = range is null
                ? await store.GetContent(key)
                : await store.GetPartialContent(key, range);

            await StreamCopyOperation.CopyToAsync(stream, httpContext.Response.Body, count: null, bufferSize: 64 * 1024, cancel: httpContext.RequestAborted);
        },
        fileLength: await store.GetSize(key)
    );

// respond with chunks representing the selected part of a resource
// the chunks might be "wasting" (in this case they have a fixed size),
// but they might come from a local memory cache
app.MapGet("/example2/{key}", async (string key, IChunkStore store) => {
    var size = await store.GetSize(key);
    return TypedResults.Stream(
        async (range, httpContext) =>
        {
            var chunkSize = store.GetChunkSize();
            var to = (range?.To + 1) ?? size;
            var start = range?.From ?? (size - to);
            var end = range?.From is null ? size : to;

            while (start < end)
            {
                var chunk = await store.GetChunk(start / chunkSize);
                var slice = chunk.AsMemory().Slice(
                    (int)(start % chunkSize),
                    (int)Math.Min(chunkSize, end - start)
                );
                await httpContext.Response.Body.WriteAsync(slice);
                start += slice.Length;
            }
        },
        fileLength: size,
    )
});
```

## Alternative Designs

The proposed API exposes the range information as `RangeItemHeaderValue?`.
There are several alternatives to represent the range; this type was proposed based on what is used internally/emitted by `HttpResultsHelper.WriteResultAsFileCore`.

The proposed API returns a `PushStreamHttpResult`, under the assumption that the same class can be used for both methods implementing responses-pushing-to-body.
This would involve changes to `PushStreamHttpResult`; another approach would be to write a separate `IResult` implementation in `Microsoft.AspNetCore.Http.HttpResults`.

As mentioned in the background, it is possible to develop an independent `IResult` implementation; this requires duplicating the functionality implemented in `HttpResultsHelper.WriteResultAsFileCore`.

Another option is to wrap this as a `Stream` and use the `TypedResults.Stream(Stream stream, ...)` overload, but this is inconvenient for two reasons:
 - this makes push-based operations harder to express (which I believe is the reason for the existence of `TypedResults.Stream(Func<Stream, Task> stream, ...)`)
 - the async creation 

See https://gist.github.com/ranma42/a528555972f16c17ed1b840bfe7fbf5c for an example of the stream hack.

## Risks

The API should be an extension of the existing API and AFAICT involves no breaking change.

I have experimented locally with an implementation that extends `PushStreamHttpResult`; in order to reuse the same class, some minor changes are needed that should not result in performance regressions.
A straightforward implementation would simply replace the callback with a `Func<RangeItemHeaderValue?, HttpContext, Task>` and wrap the `Func<Stream, Task> streamWriterCallback` as `(range, httpContext) => streamWriterCallback(httpContext.Response.Body)`.

**Comment by ranma42** [comment]
I pushed my experiment as https://github.com/ranma42/aspnetcore/tree/partial-push-stream

**Comment by ranma42** [comment]
I see that the range has probably been intentionally omitted. https://github.com/dotnet/aspnetcore/issues/39383#issuecomment-1020858817 mentions that it could be implemented manually, but I did not find any straightforward way to do that (without re-implementing the whole range computation machinery). Maybe it would be an interesting example to add in the docs.

**Comment by halter73** [comment]
Is there a particular reason you specified `HttpContext` instead of `Stream` as a parameter for the function? If you really need something else from the `HttpContext`, you're free to add it as a parameter to your handler. I think a `Stream` parameter makes more sense considering the other overload.

If we do this, we'd want to add the same overload to the untyped `Results` class too. However, once you start dealing with something as complex as manually interpreting the `RangeItemHeaderValue`, I wonder if you're better off operating on the `HttpContext` directly rather than returning an `IResult`. You should be able to get the `RangeItemHeaderValue` yourself pretty easily using [RangeHeaderValue.TryParse](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.headers.rangeheadervalue.tryparse?view=net-8.0).

Another concern is overloading based on `Func` arity. If we add this, it would be the only overload of `TypeResults.Stream` that can take a `Func` of two arguments as the first parameter without making a source breaking change. It wouldn't matter if the `Func` argument types are different because the types are often inferred for lambdas.

**Comment by ranma42** [comment]
> Is there a particular reason you specified `HttpContext` instead of `Stream` as a parameter for the function? If you really need something else from the `HttpContext`, you're free to add it as a parameter to your handler. I think a `Stream` parameter makes more sense considering the other overload.

I agree, the body `Stream` would work just fine as parameter; should I update the proposal above?

> If we do this, we'd want to add the same overload to the untyped `Results` class too.

Yup, I did that in the experiment I mentioned here https://github.com/dotnet/aspnetcore/issues/55753#issuecomment-2118637102
If desired, I can add it to the proposal (it is basically going to have the very same signature, except for the return type... which I just noticed I forgot to set to `IResult` in my experiment 😅 ).

> However, once you start dealing with something as complex as manually interpreting the `RangeItemHeaderValue`, I wonder if you're better off operating on the `HttpContext` directly rather than returning an `IResult`. You should be able to get the `RangeItemHeaderValue` yourself pretty easily using [RangeHeaderValue.TryParse](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.headers.rangeheadervalue.tryparse?view=net-8.0).

The non-trivial part that this (and other `IResult` implementations) takes care of is actually the `HttpResultsHelper.WriteResultAsFileCore` part, which is currently not exposed and performs a non-trivial amount of work around the response headers (including shortcuts to 304 (NotModified), 412 (PreconditionFailed), processing of `If-Range` and range normalization).
Even discounting that, a manual implementation would most likely have to at least re-implement the logic in `Microsoft.AspNetCore.Internal.RangeHelper`, which is internal.

> Another concern is overloading based on `Func` arity. If we add this, it would be the only overload of `TypeResults.Stream` that can take a `Func` of two arguments as the first parameter without making a source breaking change. It wouldn't matter if the `Func` argument types are different because the types are often inferred for lambdas.

I am not sure I understand the concern. IIUC this is not going to be a breaking change and the function arity difference should avoid issues with type inference 🤔 
Do you think it would be better to name the method differently (example: `PartialStream`, `RangeStream`, ...) to completely avoid adding this as an overload?

**Comment by halter73** [comment]
API Review Notes:

- > Do you think it would be better to name the method differently (example: `PartialStream`, `RangeStream`, ...) to completely avoid adding this as an overload?
  - We're weary about adding new `Stream` overloads. Too many overloads can create unnecessary confusion for the common case in order to support a niche use case. So far, we've only seen one person asking for this feature.
  - Renaming to `PartialStream` or `RangeStream` would help. It would make it a little clearer that everything written to the `Stream` would be used and there wouldn't be any built-in logic to skip over any data written that isn't part of the range unlike the current `Stream` overload that takes a seekable Stream directly.
  - However, even with a different name, we aren't sure if this use case is common enough to warrant a method on `Results`/`TypedResults`.

- `enableRangeProcessing = false` doesn't make much sense. Shouldn't `enableRangeProcessing` always be true if the func takes a `RangeItemHeaderValue`?

- `PipeWriter` would probably be more efficient than `Stream` as an argument, but we've avoided overloads for both to [avoid ambiguity](https://github.com/dotnet/aspnetcore/issues/39383#issuecomment-1020858817). Do we want to make this problem worse by introducing event more `Results` methods that take a func with a `Stream` parameter?
  - The original proposal that takes an `HttpContext` sidesteps this issue, but it's less consistent with the other `Results` methods.

- > The non-trivial part that this (and other `IResult` implementations) takes care of is actually the `HttpResultsHelper.WriteResultAsFileCore` part, which is currently not exposed and performs a non-trivial amount of work around the response headers (including shortcuts to 304 (NotModified), 412 (PreconditionFailed), processing of `If-Range` and range normalization).
  - This does explain why `Request.GetTypedHeaders().Range?.Ranges` is insufficient, but not why an API similar to `WriteResultAsFileCore` would be insufficient. For such an advanced use case, we would rather have a static API similar to `WriteResultAsFileCore` that takes an `HttpContext` as a parameter, set the appropriate response headers, and returns the normalized `RangeItemHeaderValue` if it's valid.
  - It's not clear where it would go (`PushStreamHttpResult`?) or what we would call it (`SetFileHeaders`)?
  - In the meantime (while we agree this isn't clean), your `LazyStream` workaround that reads the `ContentRange` response header to create a fake seekable stream seems fine. Ideally, it would verify the reader of the `LazyStream` actually does seek to the beginning of the `ContentRage`, but I cannot imagine why it wouldn't.

tl;dr: We don't like a new `Results`/`TypedResults` method for this, because we don't think it's a common enough scenario. If there were more people asking for this, we might reconsider. In the meantime, we would consider a sufficiently out-of-the-way static helper method if we could figure out where to put it and what to call it.

**Comment by BickelLukas** [comment]
This Api would also solve our problem. We have files stored in object storage and want to stream directly from the storage to the client to let the user download his files (with proper authorization and metadata handling in our api). 
The storage exposes api's to either get the complete file or get a partial chunk. At the moment we're able to do it by overriding `FileResult` in a Controller but this is blocking us from switching to MinimalApi since all the classes are sealed or internal.

**Comment by nickwb** [comment]
I would also like this feature. This is useful when the ASP.NET application will serve a request body that will be retrieved from an upstream source that can also support range headers and partial content.

I agree, as other have noted that:

- It would be helpful if `RangeHelper.ParseRange` was exposed as a public API
- A `TypedResult.PartialStream` would be ideal. Perhaps it could require the caller to "confirm" the range of bytes which were being served in the response.
- Alternatively, some way to utilise `FileResultsHelper.SetHeadersAndLog` or `FileResultsHelper.SetRangeHeaders` from a public API surface would be helpful.

I will point out a specific example where this would be helpful and the Microsoft team may wish to support. The [`Azure.Storage.Blobs` .NET SDK](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs#L1114) has specific support for an `HttpRange` argument when streaming a blob. However, if you don't use this overload (i.e., don't provide an `HttpRange`), the resulting `Stream` is **not** seekable. So passing the stream as an argument to `TypedResults.Stream` will cause the range processing to fail and fallback to serving the entire blob.
